docs/proposals/config_data.md

```diff
@@ -0,0 +1,316 @@
+# Generic Configuration Object
+
+## Abstract
+
+This proposal proposes a generic configuration object named `ConfigData` stored inside `etcd` that
+stores data used for the configuration of applications deployed on `Kubernetes`.
+
+The main focus points of this proposal are:
+
+* Dynamic distribution of configuration data to components and deployed applications.
+* Encapsulate configuration information and simplify `Kubernetes` deployments.
+* Create a flexible configuration model for `Kubernetes`.
+
+## Motivation
+
+Currently command-line flags and environment variables are used for configurations in `Kubernetes`
+which is complicated and error-prone. This approach is not suitable for a distributed computing environment.
+Specifically the problems with the current approach are:
+
+* Complicated deployment orchestration
+* Hard to synchronize across replicated components.
+* Hard to change dynamically as needed by continuous services.
+
+## Use Cases
+
+1. Be able to load API resources as environment variables into a container.
+2. Be able to load API resources as mounted volumes into a container.
+3. Be able to update mounted volumes after a change to the API resource.
+
+## Solution
+
+A new `ConfigData` API resource is proposed to address these problems.
+
+The `ConfigData` architecture will be very similar to that of `Secrets`.
+
+`ConfigData` can either be mounted as a volume or can be injected as environment
+variables to a container through the use of the downward API.
+
+### Dynamic Configuration
+
+Any long-running system has mutating specification over time. In order to facilitate this functionality,
+`ConfigData` will be versioned and updates will automatically made available to the container.
+
+`resourceVersion` (Found in `ObjectMetadata`) of the `ConfigData` object will be updated by the API server every time the object is modified.  After an update, modifications will be made visible to the consumer container. If the consumer uses the `Data` pairs only for initialization or during starting process, A rolling-update might be necessary to update the containers.
+
+It is then the consumer container's responsibility to make use of the updated data
```

Comment:
Leaving that responsibility to the application would be a violation of encapsulation. Application shouldn't depend on anything like that. The volume will be updated so that if an application is using the volume constantly it will use the modified volume after its updated. If the volume is used for one-time configuration of the container, this container should be updated using a kubernetes replication mechanism (rolling updates in particular) or an init container.